File: src/leetcode/p151to200/LeetCode155MinStack.java
package leetcode.p151to200;

import java.util.Stack;

/**
 * @see leetcode.p701to750.LeetCode716MaxStack
 */
public class LeetCode155MinStack {

    private interface MinStack {

        void push(int x);

        void pop();

        int top();

        int getMin();
    }

    private static class OneStack implements MinStack {

        private int mMin = Integer.MAX_VALUE;
        private Stack<Long> mStack = new Stack<>();

        @Override
        public void push(int x) {
            mStack.push((long) x - mMin);
            mMin = Math.min(mMin, x);
        }

        @Override
        public void pop() {
            if (mStack.isEmpty()) {
                return;
            }
            long num = mStack.pop();
            if (num < 0) {
                mMin -= num;
            }
        }

        @Override
        public int top() {
            long num = mStack.peek();
            if (num < 0) {
                return mMin;
            } else {
                return (int) (num + mMin);
            }
        }

        @Override
        public int getMin() {
            return mMin;
        }
    }

    private static class TwoStack implements MinStack {

        private Stack<Integer> mOriStack = new Stack<>();
        private Stack<Integer> mMinStack = new Stack<>();

        @Override
        public void push(int x) {
            mOriStack.push(x);
            if (mMinStack.isEmpty()) {
                mMinStack.push(x);
            } else {
                mMinStack.push(Math.min(mMinStack.peek(), x));
            }
        }

        @Override
        public void pop() {
            mOriStack.pop();
            mMinStack.pop();
        }

        @Override
        public int top() {
            return mOriStack.peek();
        }

        @Override
        public int getMin() {
            return mMinStack.peek();
        }
    }
}

File: src/leetcode/p151to200/LeetCode155MinStack.kt
package leetcode.p151to200

import java.util.Stack

class LeetCode155MinStack {

    interface Method {
        fun push(x: Int)
        fun pop()
        fun top(): Int
        fun getMin(): Int
    }

    class OneStask_Method : Method {

        private var min: Long = 0
        private val stack: Stack<Long> = Stack()

        override fun push(x: Int) {
            val num = x.toLong()
            if (stack.isEmpty()) {
                // !!!!!! push(0l)
                stack.push(0L)
                min = num
            } else {
                stack.push(num - min)
                if (num < min) {
                    min = num
                }
            }
        }

        override fun pop() {
            if (stack.isEmpty()) {
                return
            }
            val num = stack.pop()
            if (num < 0) {
                min -= num
            }
        }

        override fun top(): Int {
            val num = stack.peek()
            return if (num < 0) {
                min.toInt()
            } else {
                (num + min).toInt()
            }
        }

        override fun getMin(): Int {
            return min.toInt()
        }
    }

    class TwoStack_Method : Method {

        private val minStack = Stack<Int>()
        private val norStack = Stack<Int>()

        override fun push(x: Int) {
            if (norStack.isEmpty()) {
                minStack.push(x)
            } else {
                minStack.push(Math.min(x, minStack.peek()))
            }
            norStack.push(x)
        }

        override fun pop() {
            minStack.pop()
            norStack.pop()
        }

        override fun top(): Int {
            return norStack.peek()
        }

        override fun getMin(): Int {
            return minStack.peek()
        }
    }
}
